sling.go

```diff
@@ -517,17 +514,50 @@ func (s *Sling) doDecode(req *http.Request, successV, failureV interface{}) (*ht
 // decoding is skipped.
 // Caller is responsible for closing the resp.Body.
 func decodeResponse(resp *http.Response, decoder ResponseDecoder, successV, failureV interface{}) error {
-	if code := resp.StatusCode; 200 <= code && code <= 299 {
+	if code := resp.StatusCode; isSuccessful(code) {
 		if successV != nil {
 			return decoder.Decode(resp, successV)
 		}
 	} else {
 		if failureV != nil {
 			return decoder.Decode(resp, failureV)
 		}
-		var buf bytes.Buffer
-		buf.ReadFrom(resp.Body)
-		return fmt.Errorf(buf.String())
+
+		body, err := readWithCap(resp.Body, 100)
+		if err != nil {
+			return fmt.Errorf("status code %d was not successful and could not get body: %w", code, err)
+		}
+
+		return fmt.Errorf("status code %d was not successful, got body: %s", code, body)
 	}
 	return nil
 }
+
+func isSuccessful(code int) bool {
+	return 200 <= code && code <= 299
+}
+
+func readWithCap(r io.Reader, cap int) (string, error) {
```

Comment:
How about you use the io.LimitReader?

Reply:
Unfortunately that doesn't seem to actually make it nicer since the three cases:
- `len(body) < maxBodySize` - The body doesn't have to be truncated
- `len(body) == maxBodySize` - The body _still_ doesn't have to be truncated
- `len(body) > maxBodySize` - The body is truncated

Makes `len(body) >= maxBodySize` both give `io.EOF` so the code is basically the same.

This is the change we came up with in person:
```go
var maxBodySize = 100

func readWithCap(r io.Reader, cap int64) (string, error) {
	r = &io.LimitedReader{R: r, N: cap+1}

	buf, err := io.ReadAll(r)
	if err != nil && !errors.Is(err, io.ErrUnexpectedEOF) {
		return "", err
	}

	if len(buf) == 0 {
		return "", fmt.Errorf("got no response content")
	}

	var body string
	if len(buf) <= maxBodySize {
		body = string(buf)
	} else {
		// If the body is larger than the maxBodySize, it has to be truncated
		body = string(buf[:maxBodySize]) + " (truncated)"
	}

	return body, nil
}
```

Which isn't much different or much better.